agent/src/agentclient.rs

```diff
@@ -3,102 +3,134 @@ use k8s_openapi::api::core::v1::Node;
-    BottlerocketNodeStatus,
+    BottlerocketNode, BottlerocketNodeSelector, BottlerocketNodeState, BottlerocketNodeStatus,
 };
 
-const AGENT_SLEEP_DURATION: Duration = Duration::from_secs(5);
+// The reflector uses exponential backoff.
+// These values configure how long to delay between tries.
+const RETRY_BASE_DELAY: Duration = Duration::from_millis(1000);
+const RETRY_MAX_DELAY: Duration = Duration::from_secs(10);
+const NUM_RETRIES: usize = 5;
+
 const NODE_RESOURCE_ENDPOINT: &'static str = "/bottlerocket-node-resource";
+const AGENT_SLEEP_DURATION: Duration = Duration::from_secs(5);
 
 #[derive(Clone)]
 pub struct BrupopAgent {
     k8s_client: kube::client::Client,
-    node_selector: Option<BottlerocketNodeSelector>,
+    brn_reader: Store<BottlerocketNode>,
+    node_reader: Store<Node>,
+    associated_node_name: String,
+    associated_bottlerocketnode_name: String,
 }
 
 impl BrupopAgent {
-    pub fn new(k8s_client: kube::client::Client) -> Self {
+    pub fn new(
+        k8s_client: kube::client::Client,
+        brn_reader: Store<BottlerocketNode>,
+        node_reader: Store<Node>,
+        associated_node_name: String,
+        associated_bottlerocketnode_name: String,
+    ) -> Self {
         BrupopAgent {
             k8s_client,
-            node_selector: None,
+            brn_reader,
+            node_reader,
+            associated_node_name,
+            associated_bottlerocketnode_name,
         }
     }
 
-    pub async fn check_node_custom_resource_exists(&mut self) -> Result<bool> {
-        let associated_bottlerocketnode_name = node_resource_name(&self.get_node_selector().await?);
-        let bottlerocket_nodes: Api<BottlerocketNode> =
-            Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+    pub async fn check_node_custom_resource_exists(&self) -> Result<bool> {
+        // try to check if node custom resource exist in the store first. If it's not present in the
+        // store(either node custom resource doesn't exist or store data delays), make the API call for second round check.
 
-        Ok(bottlerocket_nodes
-            .get(&associated_bottlerocketnode_name)
-            .await
-            .context(error::BottlerocketNodeNotExist {
-                node_name: associated_bottlerocketnode_name,
-            })
-            .is_ok())
+        let associated_bottlerocketnode = self.brn_reader.state().clone();
+        if associated_bottlerocketnode.len() != 0 {
+            Ok(true)
+        } else {
+            let bottlerocket_nodes: Api<BottlerocketNode> =
+                Api::namespaced(self.k8s_client.clone(), NAMESPACE);
+
+            Ok(bottlerocket_nodes
+                .get(&self.associated_bottlerocketnode_name.clone())
+                .await
+                .context(error::BottlerocketNodeNotExist {
+                    node_name: &self.associated_bottlerocketnode_name.clone(),
+                })
+                .is_ok())
```

Comment:
One problem with this is that it could return `Ok(false)` under cases where the `brn` *does* exist but communication with the k8s API fails for some other reasons. I wonder if it's possible to `match` on an error variant that tells us if it doesn't exist?